### DISCUSSION TITLE
How to annotate kwargs with Unpack and TypedDict?

### DISCUSSION THREAD

**Discussion by 93578237** [python/typing#1137] [Q&A] [answered]
Consider this example:
`foo` should require either `x` or `y` is required, not both at the same time.
```python
class Kwargs(TypedDict, total=False):
    x: Required[int]
    y: Required[int]
    z: NotRequired[str]


def foo(**kwargs: Unpack[Kwargs]):
   ....
   
foo(x=1) # should be ok
foo(y=1) # should be ok
foo(z=1) # raise error

**Comment by JelleZijlstra** [comment, marked as answer]
This isn't something you can do with a TypedDict, and I don't think the upcoming PEP for TypedDict + kwargs will support it. You should use overloads instead.

**Reply by 93578237** [reply]
Hmm, but is it possible to combine overload with Unpack[TypedDict] for other NotRequired kwargs? Using only @overload makes lots of combinations.

**Reply by JelleZijlstra** [reply]
I feel like that should be OK. Perhaps try it with pyright, which already supports this feature.

**Reply by 93578237** [reply]
like this?

```python
class Kwargs(TypedDict, total=False):
    z: NotRequired[str]
    c: NotRequired[str]
    d: NotRequired[str]

class _X(Kwargs):
    x: Required[int]

class _Y(Kwargs):
    y: Required[int]


@overload
def foo(**kwargs: Unpack[_X]) -> None:
    ...


@overload
def foo(**kwargs: Unpack[_Y]) -> None:
    ...


def foo(**kwargs):
    print(kwargs)
```

**Reply by erictraut** [reply]
Yes, that's correct. This type checks as expected in pyright. In particular, it allows `foo(x=1)` and `foo(y=1)` but rejects `foo(z=1)`.

**Reply by 93578237** [reply]
Yes, pyright works, thx!
